README.md
# react-handsontable-hoc

[![npm](https://img.shields.io/npm/v/@ohtomi/react-handsontable-hoc.svg)](https://www.npmjs.com/package/@ohtomi/react-handsontable-hoc)
[![License](https://img.shields.io/npm/l/@ohtomi/react-handsontable-hoc.svg)](LICENSE)
[![Build Status](https://travis-ci.org/ohtomi/react-handsontable-hoc.svg?branch=master)](https://travis-ci.org/ohtomi/react-handsontable-hoc)
[![Greenkeeper badge](https://badges.greenkeeper.io/ohtomi/react-handsontable-hoc.svg)](https://greenkeeper.io/)

## Description

A higher order component for `react-handsontable`.

## How to use

```javascript
const data = [
    {'id': 11, 'name': 'ford', 'year': 2015, 'volume': 1000, 'processed': true},
    {'id': 12, 'name': 'ford', 'year': 2016, 'volume': 1000, 'processed': true},
    {'id': 13, 'name': 'ford', 'year': 2017, 'volume': 1000, 'processed': true},
    {'id': 14, 'name': 'ford', 'year': 2018, 'volume': 1000, 'processed': false},
    {'id': 21, 'name': 'volvo', 'year': 2015, 'volume': 1000, 'processed': true},
    {'id': 22, 'name': 'volvo', 'year': 2016, 'volume': 1000, 'processed': true},
    {'id': 23, 'name': 'volvo', 'year': 2017, 'volume': 1000, 'processed': true},
    {'id': 24, 'name': 'volvo', 'year': 2017, 'volume': 1000, 'processed': false},
    {'id': 31, 'name': 'toyota', 'year': 2016, 'volume': 1000, 'processed': true},
    {'id': 32, 'name': 'toyota', 'year': 2017, 'volume': 1000, 'processed': true},
    {'id': 33, 'name': 'toyota', 'year': 2018, 'volume': 1000, 'processed': true},
    {'id': 41, 'name': 'honda', 'year': 2015, 'volume': 1000, 'processed': true}
]

const columns = [
    {data: 'id', type: 'numeric', width: 150, readOnly: true},
    {data: 'name', type: 'text', width: 150, readOnly: true},
    {data: 'year', type: 'numeric', width: 150, readOnly: true},
    {data: 'volume', type: 'numeric', width: 150, readOnly: true},
    {data: data => data.processed ? 'Yes' : 'No', type: 'text', width: 150, readOnly: true}
]

const colHeaders = ['ID', 'NAME', 'YEAR', 'VOLUME', 'PROCESSED?']

// sort by YEAR
const initialColumnSorting = {
    initialConfig: {
        column: 2,
        sortOrder: 'desc'
    }
}

// VOLUME is hidden
const manualColumnsHide = [3]

// filter by NAME
const rowFilter = new RowFilter([
    {
        physical: 1,
        expression: Expressions.get({
            symbol: 'by_values',
            props: ['ford', 'volvo']
        })
    }
])

const afterRowSelection = (rows) => {
    // rows is an array of physical row index
}

const afterRowFiltering = (filteredRows) => {
    // filteredRows is the number of rows filtered by given rowFilter
}

const onClickColHeaderButton = (column, buttonEl) => {
    // column is a physical column index of the clicked column header button
    // buttonEl is an HTML element of the clicked column header button
}

const App = () => (
    <HotTableContainer
        data={data} columns={columns} colHeaders={colHeaders}
        width="800" height="350"
        manualColumnMove={true}
        // for Row Selection
        selectionMode="row"
        afterRowSelection={afterRowSelection}
        // for Initial Column Sorting
        columnSorting={true}
        initialColumnSorting={initialColumnSorting}
        // for Manual Column Hide
        manualColumnsHide={manualColumnsHide}
        manualColumnResize={true}
        // for Row Filter
        rowFilter={rowFilter}
        colHeaderButtonClassName={'my-col-header-button'}
        afterRowFiltering={afterRowFiltering}
        onClickColHeaderButton={onClickColHeaderButton}
    />
)

HotTablePlugins.registerPlugins()
render(<App/>, document.getElementById('root'))
```

## How to build

```bash
$ npm install
$ CI=true npm test
$ npm run build
```

## Contributing

1. Fork it!
1. Create your feature branch: `git checkout -b my-new-feature`
1. Commit your changes: `git commit -am 'Add some feature'`
1. Push to the branch: `git push origin my-new-feature`
1. Submit a pull request :D

## License

MIT

## Author

[Kenichi Ohtomi](https://github.com/ohtomi)


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## 0.6.6 (2019-03-23)

### Changed

- Upgrade `react` packages from `16.6.3` to `16.8.5`
- Upgrade `handsontable` package from `6.1.1` to `6.2.2`

### Fixed

- Use `numbro` `2.1.0` to avoid `RangeError` problem caused by `2.1.1`
  - see [Range Error after upgrade to 2.1.1](https://github.com/BenjaminVanRyseghem/numbro/issues/402)
- Remove `float` from row filter button's style and tweak its position

## 0.6.5 (2018-11-27)

### Fixed

- Remove test files from dist/

## 0.6.4 (2018-11-26)

### Fixed

- Remove test files from dist/

## 0.6.3 (2018-11-26)

### Fixed

- Remove test files from dist/

## 0.6.2 (2018-11-26)

### Fixed

- Remove test files from dist/

## 0.6.1 (2018-11-25)

### Fixed

- NPM modules will be installed when a library user installed this library
  - @handsontable/react 2.0.0
  - handsontable 6.1.1
  - shallow-equal-props 1.0.2

## 0.6.0 (2018-11-24)

### Added

- New properties of `HotTableContainer`
  - `initialColumnSorting` is to sort after loading data
  - `manualColumnsHide` is to hide columns
  - `afterRowSelection` is to notify selected physical row index
  - `afterRowFiltering` is to notify the number of filtered rows
- `Handsontable` from `handsontable` is exported to support custom renderer

### Changed

- A type of `initialColumnSorting`'s `sortOrder` is changed from `boolean` to `text`, according to `Handsontable` v6.1.1
- When column header button clicked, a physical column index will be passed
- `HotTableContainer` properties
  - from `onClickRowFilterIndicator` to `onClickColHeaderButton`
  - from `rowFilterIndicatorClassName` to `colHeaderButtonClassName`
- `HotTableContainer` is wrapped by `React.memo` to make it pure component
- Remove `reevaluate` method from `RowFilter`

### Fixed

- An unnecessary space, occupied by hidden columns, on the right side of a table is trimmed

### Removed

- `HotTableContainer` properties
  - `columnMapping`
  - `hiddenColumns`
  - `rowHeaders`

## 0.5.1 (2017-11-06)

### Added

- Generate source map

## 0.5.0 (2017-10-19)

### Fixed

- Fix initial column moving problem

### Added

- Add selection mode
  - `cell`: current or area selection (default)
  - `row`: row(s) selection

## 0.4.0 (2017-09-10)

### Fixed

- Fix column order reset bug if filtered data is empty array

## 0.3.0 (2017-09-01)

### Fixed

- Fix initial column movement bug

## 0.2.0 (2017-08-27)

### Added

- Render buttons as row filter indicator on column header if both of `rowFilter` and `onClickRowFilterIndicator` are specified
- Add `rowFilter` to filter rows
- Add `onClickRowFilterIndicator` to handle click events from row filter indicators

### Changed

- Make `hiddenColumns` updatable

### Fixed

- Fix `eslint` errors
- Fix `flowtype` errors

## 0.1.0 (2017-08-15)

Initial release

### Added

- Add Fundamental features


src/index.js
import React from 'react'
import {render} from 'react-dom'

import {Expressions, HotTableContainer, HotTablePlugins, RowFilter} from './lib'


const data = [
    {'id': 11, 'name': 'ford', 'year': 2015, 'volume': 1000, 'processed': true},
    {'id': 12, 'name': 'ford', 'year': 2016, 'volume': 1000, 'processed': true},
    {'id': 13, 'name': 'ford', 'year': 2017, 'volume': 1000, 'processed': true},
    {'id': 14, 'name': 'ford', 'year': 2018, 'volume': 1000, 'processed': false},
    {'id': 21, 'name': 'volvo', 'year': 2015, 'volume': 1000, 'processed': true},
    {'id': 22, 'name': 'volvo', 'year': 2016, 'volume': 1000, 'processed': true},
    {'id': 23, 'name': 'volvo', 'year': 2017, 'volume': 1000, 'processed': true},
    {'id': 24, 'name': 'volvo', 'year': 2017, 'volume': 1000, 'processed': false},
    {'id': 31, 'name': 'toyota', 'year': 2016, 'volume': 1000, 'processed': true},
    {'id': 32, 'name': 'toyota', 'year': 2017, 'volume': 1000, 'processed': true},
    {'id': 33, 'name': 'toyota', 'year': 2018, 'volume': 1000, 'processed': true},
    {'id': 41, 'name': 'honda', 'year': 2015, 'volume': 1000, 'processed': true}
]

const columns = [
    {data: 'id', type: 'numeric', width: 150, readOnly: true},
    {data: 'name', type: 'text', width: 150, readOnly: true},
    {data: 'year', type: 'numeric', width: 150, readOnly: true},
    {data: 'volume', type: 'numeric', width: 150, readOnly: true},
    {data: data => data.processed ? 'Yes' : 'No', type: 'text', width: 150, readOnly: true}
]

const colHeaders = ['ID', 'NAME', 'YEAR', 'VOLUME', 'PROCESSED?']

// sort by YEAR
const initialColumnSorting = {
    initialConfig: {
        column: 2,
        sortOrder: 'desc'
    }
}

// VOLUME is hidden
const manualColumnsHide = [3]

// filter by NAME
const rowFilter = new RowFilter([
    {
        physical: 1,
        expression: Expressions.get({
            symbol: 'by_values',
            props: ['ford', 'volvo']
        })
    }
])

class App extends React.Component {

    constructor(props) {
        super(props)

        this.afterRowSelection = this.afterRowSelection.bind(this)
        this.afterRowFiltering = this.afterRowFiltering.bind(this)
        this.showRowFilterPopover = this.showRowFilterPopover.bind(this)

        this.state = {
            messages: []
        }
    }

    render() {
        return (
            <div>
                <HotTableContainer
                    data={data} columns={columns} colHeaders={colHeaders}
                    width="800" height="250"
                    manualColumnMove={true}
                    // for Row Selection
                    selectionMode="row"
                    afterRowSelection={this.afterRowSelection}
                    // for Initial Column Sorting
                    columnSorting={true}
                    initialColumnSorting={initialColumnSorting}
                    // for Manual Column Hide
                    manualColumnsHide={manualColumnsHide}
                    manualColumnResize={true}
                    // for Row Filter
                    rowFilter={rowFilter}
                    colHeaderButtonClassName={'my-col-header-button'}
                    afterRowFiltering={this.afterRowFiltering}
                    onClickColHeaderButton={this.showRowFilterPopover}
                />
                <textarea rows={5} cols={100} value={this.state.messages.join('\n')} readOnly={true}/>
            </div>
        )
    }

    afterRowSelection(rows) {
        // rows is an array of physical row index
        const messages = [`afterRowSelection(${rows}) @${new Date()}`].concat(this.state.messages)
        this.setState({messages})
    }

    afterRowFiltering(filteredRows) {
        // filteredRows is the number of rows filtered by given rowFilter
        const messages = [`afterRowFiltering(${filteredRows}) @${new Date()}`].concat(this.state.messages)
        this.setState({messages})
    }

    showRowFilterPopover(column, buttonEl) {
        // column is a physical column index of the clicked column header button
        // buttonEl is an HTML element of the clicked column header button
        const messages = [`showRowFilterPopover(${column}), ${buttonEl} @${new Date()}`].concat(this.state.messages)
        this.setState({messages})
    }
}

HotTablePlugins.registerPlugins()

render(<App/>, document.getElementById('root'))
